src/github/github.py

```diff
@@ -0,0 +1,188 @@
+
+@dataclass
+class GithubTeam:
+    id: int
+    url: str
+    name: str
+    # slugs are the space-free special names that github likes to use
+    slug: str
+
+async def _github_request_get(
+        url: str,
+        token: str
+) -> Response:
+    result = requests.get(
+        url,
+        headers={
+            "Accept": "application/vnd.github+json",
+            "Authorization": f"Bearer {token}",
+            "X-GitHub-Api-Version": "2022-11-28"
+        }
+    )
+    rate_limit_remaining = int(result.headers["x-ratelimit-remaining"])
+    if rate_limit_remaining < 50:
+        raise Exception("Less than 50 api calls remaining before being rate limited, please try again later")
+
+    return result
+
+async def _github_request_post(
+        url: str,
+        token: str,
+        post_data: Any
+) -> Response:
+    result = requests.post(
+        url,
+        headers={
+            "Accept": "application/vnd.github+json",
+            "Authorization": f"Bearer {token}",
+            "X-GitHub-Api-Version": "2022-11-28"
+        },
+        data=post_data
+    )
+    rate_limit_remaining = int(result.headers["x-ratelimit-remaining"])
+    if rate_limit_remaining < 50:
+        raise Exception("Less than 50 api calls remaining before being rate limited, please try again later")
+
+    return result
+
+async def _github_request_delete(
+        url: str,
+        token: str
+) -> Response:
+    result = requests.delete(
+        url,
+        headers={
+            "Accept": "application/vnd.github+json",
+            "Authorization": f"Bearer {token}",
+            "X-GitHub-Api-Version": "2022-11-28"}
+    )
+    rate_limit_remaining = int(result.headers["x-ratelimit-remaining"])
+    if rate_limit_remaining < 50:
+        raise Exception("Less than 50 api calls remaining before being rate limited, please try again later")
+
+    return result
+
+async def _github_request_put(
+        url: str,
+        token: str,
+        put_data: Any
+) -> Response:
+    result = requests.put(
+        url,
+        headers={
+            "Accept": "application/vnd.github+json",
+            "Authorization": f"Bearer {token}",
+            "X-GitHub-Api-Version": "2022-11-28"},
+        data=put_data
+    )
+    rate_limit_remaining = int(result.headers["x-ratelimit-remaining"])
+    if rate_limit_remaining < 50:
+        raise Exception("Less than 50 api calls remaining before being rate limited, please try again later")
+
+    return result
+
+async def get_user_by_username(
+        username: str
+) -> GithubUser:
+    """
+        Takes in a Github username and returns an instance of GithubUser.
+
+        May return an empty list if no such user was found.
+    """
+    result = await _github_request_get(f"https://api.github.com/users/{username}",
+                              os.environ.get("GITHUB_TOKEN"))
+    result_json = result.json()
+    return [GithubUser(user["login"], user["id"], user["name"]) for user in [result_json]]
```

Comment:
What is `[result_json]` doing? I might suggest list(result_json) instead, unless you're making the list that contains only result_json, in which case maybe bind `user = result_json` instead.

Reply:
I see what you mean now though.

Reply:
5327538